pycbc/detector.py

```diff
@@ -270,18 +270,81 @@ def time_delay_from_detector(self, other_detector, right_ascension,
         Apply the time shift for the given detector relative to the assumed
         geocentric frame and apply the antenna patterns to the plus and cross
         polarizations.
+
+        Parameters
+        ----------
+        hp: pycbc.types.TimeSeries
+            Plus polarization of the GW
+        hc: pycbc.types.TimeSeries
+            Cross polarization of the GW
+        ra: float
+            Right ascension of source location
+        dec: float
+            Declination of source location
+        polarization: float
+            Polarization angle of the source
+        method: str, Optional
+            The method to use for projecting the polarizations into the
+            detector frame. Default is 'lal'.
+        reference_time: float, Optional
+            The time to use as, a reference for some methods of projection.
+            Used by 'constant' and 'vary_polarization' methods. Uses average
+            time if not provided.
         """
-        h_lal = lalsimulation.SimDetectorStrainREAL8TimeSeries(
-                hp.astype(np.float64).lal(), hc.astype(np.float64).lal(),
-                longitude, latitude, polarization, self.frDetector)
-        return TimeSeries(
-                h_lal.data.data, delta_t=h_lal.deltaT, epoch=h_lal.epoch,
-                dtype=np.float64, copy=False)
+        # The robust and most fefature rich method which includes
+        # time changing antenna patterns and doppler shifts due to the
+        # earth rotation and orbit
+        if method == 'lal':
+            h_lal = lalsimulation.SimDetectorStrainREAL8TimeSeries(
+                    hp.astype(np.float64).lal(), hc.astype(np.float64).lal(),
+                    ra, dec, polarization, self.frDetector)
+            ts = TimeSeries(
+                    h_lal.data.data, delta_t=h_lal.deltaT, epoch=h_lal.epoch,
+                    dtype=np.float64, copy=False)
+
+        # 'constant' assume fixed orientation relative to source over the
+        # duration of the signal, accurate for short duration signals
+        # 'fixed_polarization' applies only time changing orientation
+        # but no doppler corrections
+        elif method in ['constant', 'vary_polarization']:
+            if reference_time is not None:
+                rtime = reference_time
+            else:
+                # In many cases, one should set the reference time if using
+                # this method as we don't know where the signal is within
+                # the given time series. If not provided, we'll choose
+                # the midpoint time.
+                rtime = float(hp.end_time + hp.start_time) / 2.0
+
+            if method == 'constant':
+                time = rtime
+            elif method == 'vary_polarization':
+                if not isinstance(hp, TimeSeries):
+                    raise TypeError('Polarizations must be given as time'
+                                    ' series for this method')
```

Comment:
Change this to "Waveforms must be given..." Saying polarizations made me think that you had to provide the polarization angle as a time series. Also, shouldn't you check both hp and hc here?